ISE Mock MDM Project

isrferna opened the pull request:
This PR includes uploading the files for the `ise-mock-mdm` project, using the DevNet Exchange template as a reference.

isrferna committed:
ISE Mock MDM Project

This PR includes uploading the files for the `ise-mock-mdm`
project, using the DevNet Exchange template as a reference.

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+# ise-mock-mdm
```
Mock MDM server for ISE

isrferna replied:
Done, title updated. Thanks!

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+# ise-mock-mdm
+
+This project aims to develop a mock Mobile Device Management (MDM) system that can be integrated with Cisco Identity Services Engine (ISE). The project utilizes the FastAPI framework to simulate an MDM server, providing essential API endpoints that return structured XML responses as expected by Cisco ISE.
```
Line length 100 please

isrferna replied:
Done, line length set to 100 or less.

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+
+This project aims to develop a mock Mobile Device Management (MDM) system that can be integrated with Cisco Identity Services Engine (ISE). The project utilizes the FastAPI framework to simulate an MDM server, providing essential API endpoints that return structured XML responses as expected by Cisco ISE.
+ 
+## Use Case Description
```
Rephrased a little bit:

This code provides a simulated Mobile Device Management (MDM) system designed to integrate
seamlessly with Cisco Identity Services Engine (ISE). It is an essential tool for organizations
aiming to test and validate their network security solutions and device management policies without
incurring the costs and risks associated with a live MDM system.

The solution manages device information using a YAML file, streamlining the configuration and
updates of device attributes. It allows for querying device attributes based on various criteria,
including MAC address and compliance status, providing flexibility to test different scenarios.
A crucial feature of this solution is the normalization of MAC addresses, ensuring consistent
querying regardless of their provided format.

The mock MDM system utilizes FastAPI to create API endpoints that emulate the responses of a real
MDM system, enabling Cisco ISE to interact with it as if it were genuine. This setup is ideal for
facilitating comprehensive testing and validation processes.

isrferna replied:
Looks better, thanks!

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+
+This project was written and is maintained by the following individuals:
+
+* Chetankumar Phulpagare <cphulpag@cisco.com>
```
Add your name as well :)

isrferna replied:
Thanks! I wanted to make sure you were okay with this before doing it.

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+#### 2. Fill in the New MDM Server Form:
+
+| **Field**                             | **Description**                                                                                       |
+|---------------------------------------|-------------------------------------------------------------------------------------------------------|
```
These lines are too long. Only one row is really long, rest could be kept short.

isrferna replied:
I have just updated this table. Please let me know your thoughts.

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+
+#### 1. Access ISE Configuration:
+
+- Log in to the ISE Primary PAN node.
```
The certificate of the mock MDM server needs to be trusted by ISE.

isrferna replied:
You are correct! I have updated this section to reflect that.

chetanph commented on mdm.py:
```diff
@@ -0,0 +1,108 @@
+  </deviceList>
+</ise_api>
+"""
+    return Response(content=xml_response, media_type="application/xml")
```
Add an empty line at the end.

isrferna replied:
Sorry, I missed this. Added an empty line.

chetanph commented on README.md:
```diff
@@ -0,0 +1,232 @@
+![Alt text](./images/endpoint_mdm_information.png)
+
+By following these steps, you can effectively test the integration and functionality of your mock MDM server within a Cisco ISE environment. Adjust your policies and device configurations as needed to simulate various scenarios and validate the system's response.
+
```
Add reference: https://developer.cisco.com/docs/identity-services-engine/latest/version-3/

isrferna replied:
Done, thanks!

chetanph commented on mdm.py:
```diff
@@ -0,0 +1,108 @@
+    devices_data = load_devices()
+    matched_devices = []
+
+    # Normalize the MAC address in the query value if necessary
```
This should be a separate function, you can refactor this code to make it better organized. It can be done in a separate issue as well.

isrferna replied:
Done, I have created issue #2 to address this.

chetanph commented on mdm.py:
```diff
@@ -0,0 +1,108 @@
+        # If no querycriteria provided, return all
+        matched_devices.append((mac, attrs))
+
+    # Build XML response
```
Same here, can be organized into its own separate function. Will allow us to unit test it as well.

isrferna replied:
Issue #2 will also address this. Also, I have created issue #3 to develop the unit test cases for the mock MDM server. I will start working on this issue once we close this PR.

chetanph reviewed (commented):
Thank you so much for working on this!!!

Added some comments.

isrferna committed:
Enhance `README.md` and Format `mdm.py`

- Updated `README.md` for better clarity and completeness.
- Adjusted formatting in `mdm.py` for consistency.

chetanph commented on README.md:
```diff
@@ -0,0 +1,276 @@
+This project was written and is maintained by the following individuals:
+
+* Chetankumar Phulpagare <cphulpag@cisco.com>
+* Israel Fernandez <isrferna@cisco.com>
```
Just noticed this, let's remove the email address please

isrferna replied:
Yep, great idea! That way we'll avoid spam in the work email, LOL. I just deleted our emails.

isrferna committed:
Removed emails from `README` file

chetanph reviewed (approved):
Looks good, thank you!! 🎉 

Please squash and merge.
